File: flap/src/main/java/me/yifeiyuan/flap/FlapAdapter.java
package me.yifeiyuan.flap;

import androidx.lifecycle.LifecycleObserver;
import androidx.lifecycle.LifecycleOwner;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.List;

import me.yifeiyuan.flap.extensions.ComponentPool;

import static me.yifeiyuan.flap.Preconditions.checkNotNull;

/**
 * FlapAdapter is a flexible and powerful Adapter that makes you enjoy developing with RecyclerView.
 * <p>
 * Check these also if need :
 *
 * @author 程序亦非猿 [Follow me](<a> https://github.com/AlanCheen</a>)
 * @version 1.0
 * @see Component
 * @see ComponentPool
 * @see me.yifeiyuan.flap.extensions.DifferFlapAdapter
 * <p>
 * Flap Github: <a>https://github.com/AlanCheen/Flap</a>
 * @since 1.1
 */
@SuppressWarnings("ALL")
public class FlapAdapter extends RecyclerView.Adapter<Component> {

    @NonNull
    private final Flap flap = Flap.getDefault();

    @NonNull
    private List<?> data = new ArrayList<>();

    private LifecycleOwner lifecycleOwner;

    private boolean lifecycleEnable = true;

    private boolean useFlapItemPool = true;

    @NonNull
    @Override
    public Component onCreateViewHolder(@NonNull final ViewGroup parent, final int viewType) {
        return flap.onCreateViewHolder(LayoutInflater.from(parent.getContext()), parent, viewType);
    }

    @Override
    public final void onBindViewHolder(@NonNull final Component holder, final int position) {
        //ignore
    }

    @Override
    public void onBindViewHolder(@NonNull final Component holder, final int position, @NonNull final List<Object> payloads) {
        attachLifecycleOwnerIfNeed(holder);
        flap.onBindViewHolder(holder, position, getItem(position), payloads, this);
    }

    /**
     * Attaches the component to lifecycle if need.
     *
     * @param component The component we are going to bind.
     */
    private void attachLifecycleOwnerIfNeed(@NonNull final Component component) {
        if (lifecycleEnable && lifecycleOwner != null) {
            lifecycleOwner.getLifecycle().addObserver((LifecycleObserver) component);
        }
    }

    @Override
    public int getItemCount() {
        return getData() == null ? 0 : getData().size();
    }

    @Override
    public int getItemViewType(final int position) {
        return flap.getItemViewType(getItem(position));
    }

    @Override
    public void onAttachedToRecyclerView(@NonNull final RecyclerView recyclerView) {
        super.onAttachedToRecyclerView(recyclerView);
        if (recyclerView.getContext() instanceof LifecycleOwner && lifecycleOwner == null) {
            setLifecycleOwner((LifecycleOwner) recyclerView.getContext());
        }
        if (useFlapItemPool) {
            recyclerView.setRecycledViewPool(flap.getComponentPool());
        }
    }

    @Override
    public void onViewAttachedToWindow(@NonNull Component holder) {
        super.onViewAttachedToWindow(holder);
        flap.onViewAttachedToWindow(holder, this);
    }

    @Override
    public void onViewDetachedFromWindow(@NonNull Component holder) {
        super.onViewDetachedFromWindow(holder);
        flap.onViewDetachedFromWindow(holder, this);
    }

    @Override
    public void onViewRecycled(@NonNull final Component holder) {
        super.onViewRecycled(holder);
        flap.onViewRecycled(holder, this);
    }

    @Override
    public boolean onFailedToRecycleView(@NonNull final Component holder) {
        return flap.onFailedToRecycleView(holder, this);
    }

    public FlapAdapter setLifecycleOwner(@NonNull final LifecycleOwner lifecycleOwner) {
        checkNotNull(lifecycleOwner, "lifecycleOwner can't be null here.");
        this.lifecycleOwner = lifecycleOwner;
        return this;
    }

    public FlapAdapter setLifecycleEnable(boolean lifecycleEnable) {
        this.lifecycleEnable = lifecycleEnable;
        return this;
    }

    protected Object getItem(final int position) {
        return getData().get(position);
    }

    public FlapAdapter setData(@NonNull List<?> data) {
        checkNotNull(data, "data can't be null here.");
        this.data = data;
        return this;
    }

    public void setDataAndNotify(@NonNull List<?> data) {
        checkNotNull(data, "data can't be null here.");
        this.data = data;
        notifyDataSetChanged();
    }

    @SuppressWarnings("WeakerAccess")
    @NonNull
    public List<?> getData() {
        return data;
    }

    /**
     * Set whether use the global RecycledViewPool or not.
     * <p>
     * NOTE : Call this before you call RecyclerView.setAdapter.
     *
     * @param enable true by default
     * @return this
     */
    public FlapAdapter setUseComponentPool(final boolean enable) {
        this.useFlapItemPool = enable;
        return this;
    }

}

File: flap/src/main/java/me/yifeiyuan/flap/FlapAdapter.kt
@file:Suppress("MemberVisibilityCanBePrivate", "unused", "LeakingThis")

package me.yifeiyuan.flap

import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import me.yifeiyuan.flap.ext.SwipeDragHelper
import me.yifeiyuan.flap.widget.FlapStickyHeaders
import java.util.*

/**
 * Created by 程序亦非猿 on 2021/9/22.
 *
 * Flap Github: <a>https://github.com/AlanCheen/Flap</a>
 * @author 程序亦非猿 [Follow me](<a> https://github.com/AlanCheen</a>)
 * @since 2020/9/22
 * @since 3.0.0
 */
open class FlapAdapter(private val flap: Flap = Flap(), private val flapInitBlock: (FlapApi.() -> Unit)? = null) : RecyclerView.Adapter<Component<*>>(), SwipeDragHelper.Callback, FlapStickyHeaders, FlapApi by flap {

    companion object {
        private const val TAG = "FlapAdapter"
    }

    init {
        apply {
            flapInitBlock?.invoke(flap)
        }
    }

    private var data: MutableList<Any> = mutableListOf()

    // 暂时不需要
//    private fun getData(): List<Any> {
//        return Collections.unmodifiableList(data)
//    }

    /**
     * 在顶部添加数据
     */
    open fun addData(dataList: List<Any>) {
        data.addAll(dataList)
    }

    open fun addDataAndNotify(dataList: List<Any>, byNotifyDataSetChanged: Boolean = false) {
        data.addAll(0, dataList)
        if (byNotifyDataSetChanged) {
            notifyDataSetChanged()
        } else {
            notifyItemRangeInserted(0, dataList.size)
        }
    }

    open fun setData(newDataList: MutableList<Any>) {
        data.clear()
        data.addAll(newDataList)
    }

    open fun <T : Any> setDataAndNotify(newDataList: MutableList<T>, byNotifyDataSetChanged: Boolean = false) {
        data.clear()
        data.addAll(newDataList)
        if (byNotifyDataSetChanged) {
            notifyDataSetChanged()
        } else {
            notifyItemRangeChanged(0, newDataList.size)
        }
    }

    open fun appendData(appendDataList: MutableList<Any>) {
        data.addAll(appendDataList)
    }

    open fun appendDataAndNotify(appendDataList: MutableList<Any>, byNotifyDataSetChanged: Boolean = false) {
        data.addAll(appendDataList)
        if (byNotifyDataSetChanged) {
            notifyDataSetChanged()
        } else {
            notifyItemRangeInserted(itemCount - appendDataList.size, appendDataList.size)
        }
    }

    override fun getItemCount(): Int {
        return data.size
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): Component<*> {
        return flap.onCreateViewHolder(this, parent, viewType)
    }

    override fun onBindViewHolder(component: Component<*>, position: Int) {
        this.onBindViewHolder(component, position, mutableListOf())
    }

    override fun onBindViewHolder(
            component: Component<*>,
            position: Int,
            payloads: MutableList<Any>
    ) {
        flap.onBindViewHolder(this, getItemData(position), component, position, payloads)
    }

    open fun getItemData(position: Int): Any {
        return data[position]
    }

    override fun getItemViewType(position: Int): Int {
        return flap.getItemViewType(position, getItemData(position))
    }

    override fun getItemId(position: Int): Long {
        return flap.getItemId(position, getItemData(position))
    }

    override fun onAttachedToRecyclerView(recyclerView: RecyclerView) {
        flap.onAdapterAttachedToRecyclerView(this, recyclerView)
    }

    override fun onDetachedFromRecyclerView(recyclerView: RecyclerView) {
        flap.onAdapterDetachedFromRecyclerView(this, recyclerView)
    }

    override fun onViewRecycled(component: Component<*>) {
        flap.onViewRecycled(this, component)
    }

    override fun onFailedToRecycleView(component: Component<*>): Boolean {
        return flap.onFailedToRecycleView(this, component)
    }

    override fun onViewAttachedToWindow(component: Component<*>) {
        flap.onViewAttachedToWindow(this, component)
    }

    override fun onViewDetachedFromWindow(component: Component<*>) {
        flap.onViewDetachedFromWindow(this, component)
    }

    fun insertDataAt(index: Int, element: Any, notify: Boolean = true) {
        this.data.add(index, element)
        if (notify) {
            notifyItemInserted(index)
        }
    }

    fun removeDataAt(index: Int, notify: Boolean = true) {
        data.removeAt(index)
        if (notify) {
            notifyItemRemoved(index)
        }
    }

    fun swapData(fromPosition: Int, toPosition: Int, notify: Boolean = true) {
        Collections.swap(data, fromPosition, toPosition)
        if (notify) {
            notifyItemMoved(fromPosition, toPosition)
        }
    }

    override fun onSwiped(position: Int) {
        removeDataAt(position)
    }

    override fun onMoved(fromPosition: Int, toPosition: Int) {
        swapData(fromPosition, toPosition)
    }

//    fun setupStickyHeaderView(stickyHeader: View) {
//    }
//
//    fun teardownStickyHeaderView(stickyHeader: View) {
//    }

    var stickyHeaderHandler: ((position: Int, itemData: Any) -> Boolean)? = null
    fun setupStickyHeaderHandler(block: (position: Int, itemData: Any) -> Boolean) {
        stickyHeaderHandler = block
    }

    override fun isStickyHeader(position: Int): Boolean {
        return stickyHeaderHandler?.invoke(position, getItemData(position)) ?: false
    }
}